Mathlib/RingTheory/FractionalIdeal/LocalizedAtPrime.lean

```diff
@@ -0,0 +1,210 @@
+  exact Submodule.subset_span hx
+
+/-- The inclusion `I →ₗ[A] I.localizedAtPrime P`. -/
+def localizedAtPrime_inclusion : I →ₗ[A] I.localizedAtPrime P where
+  toFun := Set.inclusion (I.self_subset_localizedAtPrime P)
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+theorem mem_localizedAtPrime_iff_eq (x : K) : x ∈ I.localizedAtPrime P ↔
+    ∃ i : I, ∃ s : P.primeCompl, x = mk 1 s • i := by
+  refine ⟨fun h ↦ ?_, fun ⟨i, s, h⟩ ↦ ?_⟩; swap
+  · exact h ▸ (smul_mem _ (mk 1 s) <| I.self_subset_localizedAtPrime P (Subtype.coe_prop i))
+  rw [mem_localizedAtPrime_iff] at h
+  refine span_induction h (fun y yI ↦ ⟨⟨y, yI⟩, 1, by rw [mk_one, one_smul]⟩) ⟨0, 1, by simp⟩ ?_ ?_
+  · rintro _ _ ⟨i₁, s₁, rfl⟩ ⟨i₂, s₂, rfl⟩
+    use s₂ • i₁ + s₁ • i₂, s₁ * s₂
+    simp only [AddSubmonoid.coe_add, coe_toAddSubmonoid, coe_smul_of_tower, smul_add]
+    congr 1
+    · rw [Submonoid.mk_smul, algebra_compatible_smul Aₚ (s₂ : A) (i₁ : K), ← mk_one_eq_algebraMap,
+        ← smul_assoc, smul_eq_mul, mk_mul, mul_one, ← mk_mul, mk_self, mul_one]
+    · rw [Submonoid.mk_smul, algebra_compatible_smul Aₚ (s₁ : A) (i₂ : K), ← mk_one_eq_algebraMap,
+        ← smul_assoc, smul_eq_mul, mk_mul, mul_comm, mul_assoc, ← mk_mul, mk_self, one_mul, mul_one]
+  · rintro y _ ⟨i, s, rfl⟩
+    refine Localization.induction_on y (fun ⟨a, s₁⟩ ↦ ⟨a • i, s₁ * s, ?_⟩)
+    rw [coe_smul_of_tower, ← smul_assoc, smul_comm, ← smul_assoc]
+    nth_rewrite 1 [← mul_one a, ← smul_eq_mul, ← smul_mk]
+    rw [smul_assoc a, smul_eq_mul, mk_mul, one_mul]
+
+variable {I} {P}
+
+theorem exists_smul_mem_of_mem_localizedAtPrime {x : K} (hx : x ∈ I.localizedAtPrime P) :
+    ∃ s ∈ P.primeCompl, s • x ∈ I := by
+  obtain ⟨i, s, rfl⟩ := (I.mem_localizedAtPrime_iff_eq P x).1 hx
+  use s, Subtype.coe_prop s
+  rw [← smul_assoc, smul_mk, mk_eq_mk', smul_eq_mul, mk'_mul_cancel_left, _root_.map_one, one_smul]
+  apply SetLike.coe_mem
+
+-- `I.localizedAtPrime P` is actually the localization of `I` at `P`.
+instance : IsLocalizedModule P.primeCompl (I.localizedAtPrime_inclusion P) where
+  map_units s := by
+    refine (End_isUnit_iff _).mpr ⟨fun ⟨_, _⟩ ⟨_, _⟩ h ↦ ?_, fun x ↦ ⟨(mk 1 s) • x, ?_⟩⟩
+    · have s0 : (s : A) ≠ 0 := fun hs ↦ (hs ▸ Subtype.coe_prop s) P.zero_mem
+      rw [algebraMap_end_apply, algebraMap_end_apply, SetLike.mk_smul_of_tower_mk,
+        Subtype.mk.injEq] at h
+      exact Subtype.ext_iff_val.mpr <| smul_right_injective K s0 h
+    · rw [algebraMap_end_apply, ← smul_assoc, smul_mk]
+      simp
+  surj' x :=
+    have ⟨s, sP, sxI⟩ := exists_smul_mem_of_mem_localizedAtPrime (Subtype.coe_prop x)
+    ⟨⟨⟨s • x, sxI⟩, ⟨s, sP⟩⟩, by rw [localizedAtPrime_inclusion]; rfl⟩
+  exists_of_eq h := ⟨1, by simpa [localizedAtPrime_inclusion] using h⟩
+
+variable (P)
+
+theorem localizedAtPrime_ne_zero (hI : I ≠ 0) : I.localizedAtPrime P ≠ 0 := by
+  simp_rw [ne_eq, eq_zero_iff, not_forall] at hI ⊢
+  rcases hI with ⟨x, xI, x0⟩
+  exact ⟨x, I.self_subset_localizedAtPrime P xI, x0⟩
```

Comment:
Here's a strengthening of the result (found by noticing that the proof of `localizedAtPrime_ne_zero` starts by doing the contrapositive).
```suggestion
theorem localizedAtPrime_eq_zero_iff : I.localizedAtPrime P = 0 ↔ I = 0 := by
  constructor
  · rw [eq_zero_iff, eq_zero_iff]
    intro h x xI
    exact h _ (self_subset_localizedAtPrime _ _ xI)
  · rintro rfl
    simp

variable (P)

theorem localizedAtPrime_ne_zero (hI : I ≠ 0) : I.localizedAtPrime P ≠ 0 :=
  mt localizedAtPrime_eq_zero_iff.mp hI
```